cpp/src/copying/gather.cu

```diff
@@ -51,224 +79,102 @@ struct bounds_checker {
- * @brief Conditionally gathers the bits of a validity bitmask.
- *
- * Gathers the bits of a validity bitmask according to a gather map.
- * If `pred(stencil[i])` evaluates to true, then bit `i` in `destination_mask`
- * will equal bit `gather_map[i]` from the `source_mask`.
- *
- * If `pred(stencil[i])` evaluates to false, then bit `i` in `destination_mask`
- * will be set to 0.
- *
- * If any value appears in `gather_map` more than once, the result is undefined.
- *
- * If any of the range [source_mask, source_mask + num_source_rows) overlaps
- * [destination_mask, destination_mask + num_destination_rows), the result is
- * undefined.
- *
- * @tparam T The type of the stencil array
- * @tparam P The type of the predicate
- * @param[in] source_mask The mask whose bits will be gathered
- * @param[in] num_source_rows The number of bits in the source_mask
- * @param[out] destination_mask The output after gathering the input
- * @param[in] num_destination_rows The number of bits in the
- * destination_mask
- * @param[in] gather_map The map that indicates where elements from the
- * input will be gathered to in the output. Length must be equal to
- * `num_destination_rows`.
- * @param[in] stencil An array of values that will be evaluated by the
- * predicate. Length must be equal to `num_destination_rows`.
- * @param[in] pred Unary predicate applied to the stencil values
- *---------------------------------------------------------------------------**/
-template <typename T, typename P>
-__global__ void gather_bitmask_if_kernel(
-    gdf_valid_type const* const __restrict__ source_mask,
-    gdf_size_type const num_source_rows, gdf_valid_type* const destination_mask,
-    gdf_size_type const num_destination_rows, gdf_index_type const* gather_map,
-    T const* stencil, P pred) {
-  using MaskType = uint32_t;
-  constexpr uint32_t BITS_PER_MASK{sizeof(MaskType) * 8};
-
-  // TODO: Update to use new bit_mask_t
-  MaskType* const __restrict__ destination_mask32 =
-      reinterpret_cast<MaskType*>(destination_mask);
-
-  gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
-
-  auto active_threads =
-      __ballot_sync(0xffffffff, destination_row < num_destination_rows);
-
-  while (destination_row < num_destination_rows) {
-    bool source_bit_is_valid{false};
-    bool const predicate_is_true{pred(stencil[destination_row])};
-    if (predicate_is_true) {
-      // If the predicate for `destination_row` is false, it's valid for
-      // `gather_map[destination_row]` to be out of bounds,
-      // therefore, only use it if the predicate evaluates to true
-      source_bit_is_valid =
-          gdf_is_valid(source_mask, gather_map[destination_row]);
-    }
-
-    bool const destination_bit_is_valid{
-        gdf_is_valid(destination_mask, destination_row)};
+using CountType = gdf_size_type;
+template <class BitType, int lane = 0>
+__device__ __inline__ CountType single_lane_reduce(BitType f) {
```

Comment:
Surely there is a more understandable name than `f`. Also, this function needs documentation.